Mathlib/Algebra/Field/Defs.lean

```diff
@@ -0,0 +1,244 @@
+/-
+Copyright (c) 2014 Robert Lewis. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Robert Lewis, Leonardo de Moura, Johannes Hölzl, Mario Carneiro
+-/
+
+import Mathlib.Algebra.Ring.Defs
+import Mathlib.Tactic.Convert
+import Std.Data.Rat
+
+/-!
+# Division (semi)rings and (semi)fields
+
+This file introduces fields and division rings (also known as skewfields) and proves some basic
+statements about them. For a more extensive theory of fields, see the `field_theory` folder.
+
+## Main definitions
+
+* `division_semiring`: Nontrivial semiring with multiplicative inverses for nonzero elements.
+* `division_ring`: : Nontrivial ring with multiplicative inverses for nonzero elements.
+* `semifield`: Commutative division semiring.
+* `field`: Commutative division ring.
+* `is_field`: Predicate on a (semi)ring that it is a (semi)field, i.e. that the multiplication is
+  commutative, that it has more than one element and that all non-zero elements have a
+  multiplicative inverse. In contrast to `field`, which contains the data of a function associating
+  to an element of the field its multiplicative inverse, this predicate only assumes the existence
+  and can therefore more easily be used to e.g. transfer along ring isomorphisms.
+
+## Implementation details
+
+By convention `0⁻¹ = 0` in a field or division ring. This is due to the fact that working with total
+functions has the advantage of not constantly having to check that `x ≠ 0` when writing `x⁻¹`. With
+this convention in place, some statements like `(a + b) * c⁻¹ = a * c⁻¹ + b * c⁻¹` still remain
+true, while others like the defining property `a * a⁻¹ = 1` need the assumption `a ≠ 0`. If you are
+a beginner in using Lean and are confused by that, you can read more about why this convention is
+taken in Kevin Buzzard's
+[blogpost](https://xenaproject.wordpress.com/2020/07/05/division-by-zero-in-type-theory-a-faq/)
+
+A division ring or field is an example of a `group_with_zero`. If you cannot find
+a division ring / field lemma that does not involve `+`, you can try looking for
+a `group_with_zero` lemma instead.
+
+## Tags
+
+field, division ring, skew field, skew-field, skewfield
+-/
+
+
+open Function Set
+
+universe u
+
+variable {α β K : Type _}
+
+/-- The default definition of the coercion `(↑(a : ℚ) : K)` for a division ring `K`
+is defined as `(a / b : K) = (a : K) * (b : K)⁻¹`.
+Use `coe` instead of `rat.cast_rec` for better definitional behaviour.
+-/
+def Rat.castRec [HasLiftT ℕ K] [HasLiftT ℤ K] [Mul K] [Inv K] : ℚ → K
```

Comment:
I'm getting stuck on the lack of `HasLiftT` already - would appreciate a pointer to how this should look now

Reply:
It seems that we are lacking a port of https://github.com/leanprover-community/lean3port/blob/master/Leanbin/Init/Coe.lean - some parts of that file are obsolete though

Reply:
I've ported this replacing `has_lift_t` with `CoeTail`. Unless someone knows a reason we really need to reproduce the `has_lift_t` distinction, this seems to work locally.